packages/alfa-css/src/grammars/color.ts

```diff
@@ -99,13 +99,78 @@ function rgbaColor(stream: Stream<Token>): Color {
   return color;
 }
 
+function hexColor(stream: Stream<Token>): Color {
+  const ident = stream.next();
+
+  if (ident === null || ident.type !== TokenType.Ident) {
+    return Transparent;
+  }
+  const hex = ident.value;
+
+  let repetitions;
+  let hasAlpha;
+
+  switch (hex.length) {
+    case 3:
+      repetitions = 2;
+      hasAlpha = false;
+      break;
+    case 4:
+      repetitions = 2;
+      hasAlpha = true;
+      break;
+    case 6:
+      repetitions = 1;
+      hasAlpha = false;
+      break;
+    case 8:
+      repetitions = 1;
+      hasAlpha = true;
+      break;
+    default:
+      return Transparent;
+  }
+
+  const bytes = [];
+  for (let i = 0, n = hex.length; i < n; i++) {
+    const byte = hex.charCodeAt(i);
+
+    if (Lang.isNumeric(byte)) {
+      bytes[i] = byte - Lang.Char.DigitZero;
+    } else if (
+      Lang.isBetween(byte, Lang.Char.SmallLetterA, Lang.Char.SmallLetterF)
+    ) {
+      bytes[i] = byte - Lang.Char.SmallLetterA + 10;
+    } else if (
+      Lang.isBetween(byte, Lang.Char.CapitalLetterA, Lang.Char.CapitalLetterF)
+    ) {
+      bytes[i] = byte - Lang.Char.CapitalLetterA + 10;
+    }
+  }
+
+  if (repetitions === 1) {
+    return {
+      red: bytes[0] * 16 + bytes[1],
+      green: bytes[2] * 16 + bytes[3],
+      blue: bytes[4] * 16 + bytes[5],
+      alpha: hasAlpha ? (bytes[6] * 16 + bytes[7]) / 255 : 1
+    };
+  }
+
+  return {
+    red: bytes[0] * 16 + bytes[0],
+    green: bytes[1] * 16 + bytes[1],
+    blue: bytes[2] * 16 + bytes[2],
+    alpha: hasAlpha ? (bytes[3] * 16 + bytes[3]) / 255 : 1
+  };
```

Comment:
The above is close! You should however not be allocating a byte array and pushing the individual bytes into it; that's way more complex than things need to be. Allocating two variables for containing the final hex and alpha value is enough and then you need two loops: One for the hex value and one for the alpha value. For each encountered byte you then append it to the associated value (hex or alpha) as many times as specified by the repetition count.